src/typval.c

```diff
@@ -2067,19 +2066,21 @@ tv_equal(
     if (tv1->v_type != tv2->v_type
 	    && ((tv1->v_type != VAR_BOOL && tv1->v_type != VAR_SPECIAL)
 		|| (tv2->v_type != VAR_BOOL && tv2->v_type != VAR_SPECIAL)))
+    {
 	return FALSE;
+    }
 
     switch (tv1->v_type)
     {
 	case VAR_LIST:
 	    ++recursive_cnt;
-	    r = list_equal(tv1->vval.v_list, tv2->vval.v_list, ic, TRUE);
+	    r = list_equal(tv1->vval.v_list, tv2->vval.v_list, ic);
 	    --recursive_cnt;
 	    return r;
 
 	case VAR_DICT:
 	    ++recursive_cnt;
-	    r = dict_equal(tv1->vval.v_dict, tv2->vval.v_dict, ic, TRUE);
```

Comment:
Here `list_equal()` and `dict_equal()` will recursively call `tv_equal()`, so `recursive` variables are TRUE.